core/state/statedb.go

```diff
@@ -765,7 +769,18 @@ func (db *DB) Commit(deleteEmptyObjects bool) (root common.Hash, err error) {
 		if obj := db.stateObjects[addr]; !obj.deleted {
 			// Write any contract code associated with the state object
 			if obj.code != nil && obj.dirtyCode {
-				db.db.TrieDB().InsertBlob(common.BytesToHash(obj.CodeHash()), obj.code)
+				// for validators - clear the code if asked to
+				// this is better than deleting the state object because recovery then is difficult
+				// this way we save space while retaining functionality
+				// IMPORTANT - note the order here for short-circuiting must not be changed
+				// otherwise TestSnapshot2 will fail because obj.IsValidator
+				// will cache the result in obj.originStorage
+				if clearValidatorWrappers && obj.IsValidator(db.db) {
+					// emptyCode a misnomer, should be emptyCodeHash
+					db.db.TrieDB().InsertBlob(emptyCode, nil)
```

Comment:
This doesn't seem right. Inserting nil at emptyCodeHash is just nil'ing the trie node at emptyCodeHash. It's not deleting the validatorWrapper's data, right?